zkevm-circuits/src/evm_circuit/execution/add_sub.rs

```diff
@@ -50,9 +50,18 @@ impl<F: Field> ExecutionGadget<F> for AddSubGadget<F> {
 
         // ADD: Pop a and b from the stack, push c on the stack
         // SUB: Pop c and b from the stack, push a on the stack
-        cb.stack_pop(select::expr(is_sub.expr().0, c.expr(), a.expr()));
-        cb.stack_pop(b.expr());
-        cb.stack_push(select::expr(is_sub.expr().0, a.expr(), c.expr()));
+
+        cb.stack_pop(Word::select(
+            is_sub.expr().0,
+            c.expr().to_word(),
+            a.expr().to_word(),
+        ));
+        cb.stack_pop(b.expr().to_word());
+        cb.stack_push(Word::select(
+            is_sub.expr().0,
+            a.expr().to_word(),
+            c.expr().to_word(),
+        ));
```

Comment:
This opcode does POP on a and b, and PUSH on c.
This means that `a` and `b` are read and `c` is written.
The StateCircuit guarantees that any read is consistent with the previous write.  From this we can optimize out the range checks on the reads, and only leave the range checks on the writes!

So we can directly read `a` and `b` from the stack and assume their corresponding `lo,hi` are 128 bits each.  From this we create `c` and prove that `c.lo` and `c.hi` are 128 bit each, and then write `c` to the stack.

In this case the native format for the `AddWordsGadget` is `lo, hi` so I think there's no need to express `a` and `b` as 8 bit limbs at any time.  `AddWordsGadget` doesn't do any range check on `sum.lo, sum.hi` so those ones do need to be split into limbs to guarantee that `sum.lo, sum.hi` are 128 bit each.

Note that this is an optimization, so we may apply it in the future.  I see that you also implemented `cb.query_word` so, so this is just a comment :)

Reply:
Hi @ed255 This is a good consolidation for when should do range check! And it definitely can be applied on other lookup with tag, since all value should be checked during write.

Just have small add-on, this `add-sub.rs` is pretty typical that we can just assume value pop from stack it's low/hi is within 128 bits and fit optimisation! But there are some other edge cases, for example value pop from stack must be an address, therefore at least 20 bytes. I applied partial optimisation on address type as well in latest commits. We can apply same tricks on other type, such as U64 range etc.